Library/OcMainLib/OpenCoreUefiUnloadDrivers.c

```diff
@@ -0,0 +1,414 @@
+                  gImageHandle,
+                  NULL,
+                  EFI_OPEN_PROTOCOL_GET_PROTOCOL
+                  );
+  if (!EFI_ERROR (Status)) {
+    BestLang = GetBestLanguageForDriver (CompNameStruct->SupportedLanguages, Language, FALSE);
+    Status   = CompNameStruct->GetDriverName (CompNameStruct, BestLang, &RetVal);
+    if (BestLang != NULL) {
+      FreePool (BestLang);
+    }
+
+    if (!EFI_ERROR (Status)) {
+      return (RetVal);
+    }
+  }
+
+  return (NULL);
+}
+
+VOID
+ReportImageName (
+  IN VOID          *Context,
+  IN CONST CHAR16  *Name,
+  IN EFI_HANDLE    Handle
+  )
+{
+  DRIVER_REPORT_CONTEXT  *ReportContext;
+
+  ReportContext = Context;
+
+  OcAsciiPrintBuffer (
+    &ReportContext->FileBuffer,
+    &ReportContext->FileBufferSize,
+    "%s\n",
+    Name
+    );
+}
+
+VOID
+UnloadImageByName (
+  IN VOID          *Context,
+  IN CONST CHAR16  *Name,
+  IN EFI_HANDLE    Handle
+  )
+{
+  EFI_STATUS            Status;
+  UINTN                 Index;
+  UNLOAD_IMAGE_CONTEXT  *UnloadContext;
+
+  UnloadContext = Context;
+
+  for (Index = 0; Index < UnloadContext->UnloadNameCount; Index++) {
+    if (StrCmp (UnloadContext->UnloadInfo[Index].UnloadName, Name) == 0) {
+      Status = gBS->UnloadImage (Handle);
+      DEBUG ((
+        EFI_ERROR (Status) ? DEBUG_WARN : DEBUG_INFO,
+        "OC: UnloadImage %s - %r\n",
+        UnloadContext->UnloadInfo[Index].UnloadName,
+        Status
+        ));
+      UnloadContext->UnloadInfo[Index].Unloaded = TRUE; ///< Do not report as 'not found' even if we failed to unload it.
+      break;
+    }
+  }
+}
+
+//
+// Perform action for all driver binding protocols which are
+// present on the same handle as loaded image protocol and
+// where the handle has a usable name.
+//
+VOID
+ProcessAllDrivers (
+  IN VOID        *Context,
+  PROCESS_IMAGE  ProcessImage
+  )
+{
+  EFI_STATUS                 Status;
+  UINTN                      HandleCount;
+  EFI_HANDLE                 *HandleBuffer;
+  EFI_LOADED_IMAGE_PROTOCOL  *LoadedImage;
+  UINTN                      Index;
+  CHAR16                     *FileName;
+  CONST CHAR16               *Name;
+
+  Status = gBS->LocateHandleBuffer (
+                  ByProtocol,
+                  &gEfiDriverBindingProtocolGuid,
+                  NULL,
+                  &HandleCount,
+                  &HandleBuffer
+                  );
+  if (EFI_ERROR (Status)) {
+    return;
+  }
+
+  for (Index = 0; Index < HandleCount; Index++) {
+    Status = gBS->HandleProtocol (
+                    HandleBuffer[Index],
+                    &gEfiLoadedImageProtocolGuid,
+                    (VOID **)&LoadedImage
+                    );
+    if (!EFI_ERROR (Status)) {
+      FileName = NULL;
+      Name     = GetStringNameFromHandle (HandleBuffer[Index], NULL); ///< Do not free this one.
+      if (Name == NULL) {
+        FileName = FindLoadedImageFileName (LoadedImage);
+        Name     = FileName;
+      }
```

Comment:
The choice of driver name here (to list in SysReport and use as the name to unload by) was prompted by the cMP5,1 (and I believe other similar era macs), where the long English component name is the only name available.

It is too slow to check every possible name, so we have to give priority to one or other name. I am currently giving the long English name priority, as I thought that was available everywhere - but it turns out that on many systems, the long component name is available of most but not all drivers.

Therefore it might make more sense to use the short image file name (which typically corresponds to the driver name, e.g. `HttpDxe`) as 'the' name, only falling back to the long name on systems where the driver name isn't available.